Add dependency coordinates for Lombok

mmoayyed opened the pull request:
Per earlier discussions on #41

SavvasMisaghMoayyed committed:
add dependency coordinates for lombok

franzbecker reviewed (commented):
Thanks for the contribution! Could you please add a test that changes the coordinates and verifies that it has an impact? Thanks!

SavvasMisaghMoayyed committed:
updated tests to check for lombok coordindate

mmoayyed commented:
Sure. I added a new class with invalid coordinates to verify build failures. I mostly wanted to avoid modifying the buildFile shared by other tests to minimize impact, but I dont exactly an expert on Spock; if there is a better way to do this, please let me know and I'd be happy to massage the config.

Also fixed a few typos in other test classes.

Thank you for the review!

SavvasMisaghMoayyed committed:
updated tests to check for lombok coordindate

franzbecker commented on src/test/groovy/io/franzbecker/gradle/lombok/LombokPluginIntegrationTest.groovy:
```diff
@@ -3,7 +3,7 @@ package io.franzbecker.gradle.lombok
 
 /**
- * Integration tests for {@link LombokPlugin}.
+ * Integration tests for {@link LombokPluginIntegrationTest}.
```
That was actually correct before ;)

franzbecker commented on src/test/groovy/io/franzbecker/gradle/lombok/LombokPluginSpec.groovy:
```diff
@@ -4,7 +4,7 @@ import io.franzbecker.gradle.lombok.task.InstallLombokTask
 import nebula.test.PluginProjectSpec
 /**
- * Unit tests for {@link LombokPlugin}.
+ * Unit tests for {@link LombokPluginSpec}.
```
That was actually correct as well ;)

franzbecker commented on src/test/groovy/io/franzbecker/gradle/lombok/LombokPluginExtensionTest.groovy:
```diff
@@ -0,0 +1,31 @@
+        runBuild('test')
+
+        then: "build fails to find the lombok dependency with bad coordinates"
+        thrown(InvalidPluginMetadataException.class)
```
That exception just tells us that 

> the plugin under test metadata cannot be read

(see https://docs.gradle.org/current/javadoc/org/gradle/testkit/runner/InvalidPluginMetadataException.html).
This could have various reasons and maybe not just that the coordinates are bad. Could you refine the assertions a bit?

mmoayyed replied:
I am not entirely sure how to do this; The build result is simply null here and I simply get back the following:

```
Caused by: org.gradle.testkit.runner.InvalidPluginMetadataException: Test runtime classpath does not contain plugin metadata file 'plugin-under-test-metadata.properties'
	at org.gradle.testkit.runner.internal.PluginUnderTestMetadataReading.readImplementationClasspath(PluginUnderTestMetadataReading.java:44)
	at org.gradle.testkit.runner.internal.PluginUnderTestMetadataReading.readImplementationClasspath(PluginUnderTestMetadataReading.java:37)
	at org.gradle.testkit.runner.internal.DefaultGradleRunner.withPluginClasspath(DefaultGradleRunner.java:156)
	at io.franzbecker.gradle.lombok.AbstractIntegrationTest.createGradleRunner(AbstractIntegrationTest.groovy:47)
	at io.franzbecker.gradle.lombok.AbstractIntegrationTest.runBuild(AbstractIntegrationTest.groovy:54)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at io.franzbecker.gradle.lombok.LombokPluginExtensionTest.Fails to find the lombok dependency if coordinates are invalid(LombokPluginExtensionTest.groovy:26)
```

Could you suggest how the assertions may be improved here?

franzbecker commented on README.md:
```diff
@@ -27,6 +27,7 @@ After applying the plugin, the Lombok annotations can be used directly in any Ja
     lombok { // optional: values below are the defaults
         version = "1.16.20"
         sha256 = "c5178b18caaa1a15e17b99ba5e4023d2de2ebc18b58cde0f5a04ca4b31c10e6d"
+        coordinates = "org.projectlombok:lombok"
```
Previously I just scrolled above it and thought this was probably good enough.
But changing the coordinates might break the `VerifyLombokTask` since the JAR is retrieved from the dependencies like this: https://github.com/franzbecker/gradle-lombok/blob/72616b354edad936c54e26033ee1f7934793de48/src/main/groovy/io/franzbecker/gradle/lombok/task/VerifyLombokTask.groovy#L44

We might need to tweak that part as well and write a test for that as well.

mmoayyed replied:
Sure. Could you please add a few examples to demonstrate how the mocked Configuration object can be used to verify the lombok dependency today? I see tests that verify the hash value of a text file, but I dont see how the lombok dependency itself is verified for integrity.

franzbecker reviewed (commented):
Thanks for the good progress, I added a few more comments. The integration test is a good step, but maybe we good add a unit test as well right below this one: https://github.com/franzbecker/gradle-lombok/blob/72616b354edad936c54e26033ee1f7934793de48/src/test/groovy/io/franzbecker/gradle/lombok/LombokPluginSpec.groovy#L49
We would need to modify the extension after applying and before evaluating the project before - don't have a proper code example but might have some time on Friday to look into it.

SavvasMisaghMoayyed committed:
fix references for unit tests in javadocs

SavvasMisaghMoayyed committed:
added unit test to verify lombok coordinates

franzbecker commented:
@mmoayyed it looks like the build is still failing :-/

mmoayyed commented:
Sorry about the delay this one. I have backed off of this change for now. My original intention for doing this was to experiment with custom lombok extensions whereby the lombok project would be forked. I have since realized that the task is simply folly and not worth the time, and so I dont plan to pursue this.

Thank you for your time, as always. If you wish to continue to working on this, please do or else, feel free to close this pull request.
